babel/dates.py

```diff
@@ -1161,7 +1166,15 @@ def parse_date(string, locale=LC_TIME, format='medium'):
     if not numbers:
         raise ParseError("No numbers were found in input")
 
-    # TODO: try ISO format first?
+    # we try ISO-8601 format first, meaning similar to formats
+    # extended YYYY-MM-DD or basic YYYYMMDD
+    iso_alike = re.match(r'^(\d{4})-?([01]\d)-?([0-3]\d)$', string)
```

Comment:
This pattern matches string with non-ascii characters. I think we should accept only-ascii string by adding `re.ASCII` or using `[0-9]` instead of `\d`.

```pycon
>>> string = '২০২৩-1২-3১'
>>> re.match(r'^(\d{4})-?([01]\d)-?([0-3]\d)$', string)
<re.Match object; span=(0, 10), match='২০২৩-1২-3১'>
```

Reply:
I added re.ASCII and rebased/squashed the whole PR.